### ISSUE TITLE
About the Roadmap to build UniTok

### ISSUE THREAD

**Issue by codeview3d** [FoundationVision/UniTok#6]
Hi, it is a really good work. After reading your paper, I have following questions : 
1. about the Roadmap to build UniTok:  In Figure 3, the CLIP baseline and +Factorization are both continous features, but + Discretization and + reconstruction can obtain the discrete tokens. In your experiments, how to deal with this? I means that may be the main performance drop comes from the continuous features to discrete features?

2. How to understand " unified multimodal model" ?  In your paper, you mention that "We proceed to develop a unified multimodal model with
UniTok. Particularly, we leverage the unified framework introduced in Liquid (Wu et al., 2024b), which models (discrete-valued) vision and language sequences with a universal next-token prediction loss". However, in your evaluate part, you use  llava( https://github.com/FoundationVision/UniTok/tree/main/eval/understanding/llava) to conduct understanding task. And Liquid (https://github.com/FoundationVision/UniTok/tree/main/eval/generation) for generation task. In general, we call unified mltimodal model means that a model to conduct understanding and generation tasks.

**Comment by machuofan** [comment]
1. As mentioned in the paper, we use the LLaVA framework for evaluation in Figure 3. The LLaVA framework consists of an LLM, a vision encoder, and an MLP connecting both. In our experiments, we replace the the vision encoder to test understanding performance of different variants of visual tokenizers. **Whether the visual token is continuous or discrete does not make a difference here**. That is, we treat discrete tokens the same as continuous tokens, which are directly projected to the language token space by the MLP. As for the performance drop caused by discretization, we have detailed analysis of this part in Section 3.2. According to our ablation, both factorization and discretization play an important role in understanding performance degradation.

2. **As mentioned in the README.md, the weights of our unified MLLM (UniTok + Liquid) will be released soon.** The currently open-sourced evaluation codes are for the LLaVA and LLamaGen (not Liquid) frameworks.

**Comment by codeview3d** [comment]
> 1. As mentioned in the paper, we use the LLaVA framework for evaluation in Figure 3. The LLaVA framework consists of an LLM, a vision encoder, and an MLP connecting both. In our experiments, we replace the the vision encoder to test understanding performance of different variants of visual tokenizers. **Whether the visual token is continuous or discrete does not make a difference here**. That is, we treat discrete tokens the same as continuous tokens, which are directly projected to the language token space by the MLP. As for the performance drop caused by discretization, we have detailed analysis of this part in Section 3.2. According to our ablation, both factorization and discretization play an important role in understanding performance degradation.
> 2. **As mentioned in the README.md, the weights of our unified MLLM (UniTok + Liquid) will be released soon.** The currently open-sourced evaluation codes are for the LLaVA and LLamaGen (not Liquid) frameworks.

OK, Thank you for your reply.

**Comment by codeview3d** [comment]
> > 1. As mentioned in the paper, we use the LLaVA framework for evaluation in Figure 3. The LLaVA framework consists of an LLM, a vision encoder, and an MLP connecting both. In our experiments, we replace the the vision encoder to test understanding performance of different variants of visual tokenizers. **Whether the visual token is continuous or discrete does not make a difference here**. That is, we treat discrete tokens the same as continuous tokens, which are directly projected to the language token space by the MLP. As for the performance drop caused by discretization, we have detailed analysis of this part in Section 3.2. According to our ablation, both factorization and discretization play an important role in understanding performance degradation.
> > 2. **As mentioned in the README.md, the weights of our unified MLLM (UniTok + Liquid) will be released soon.** The currently open-sourced evaluation codes are for the LLaVA and LLamaGen (not Liquid) frameworks.
> 
> OK, Thank you for your reply.

Expecting your unified MLLM weights release. By the way, do you have plan to release the training process of MLLM?

**Comment by machuofan** [comment]
We are seeking for the internal approval to release the training code of unified MLLM. Before that, you may refer to the training of [Liquid](https://github.com/FoundationVision/Liquid), which will open source soon. Our MLLM training codes are largely inherited from Liquid.

**Comment by yljblues** [comment]
Thank you for the great work! Also looking forward to the unified MLLM weights release!
